theurgy.lic

```diff
@@ -1,326 +1,591 @@
+    complete_or_interrupt_research
+    prepare?('bless', 1)
     forage? 'limb'
     get_from_container(@water_holder)
-    fput("sprinkle #{@water_holder} on my limb")
+    bput("sprinkle #{@water_holder} on my limb", 'You sprinkle')
     put_in_container(@water_holder)
     pause 1
-    fput 'cast my limb'
-    if !@hold_research.nil? && Flags['research_done'] == false
-      fput "RESEARCH #{@hold_research.upcase} 300"
-    end
+    cast?('cast my limb')
+    continue_research
     bput('get carving knife', 'You get', 'I could not find', 'What were you referring to')
     while right_hand.include? 'limb'
-      fput('carve my limb with my knife')
-      pause 1
+      bput('carve my limb with my knife',
+           'You begin to hack', 'With fluid strokes, you', 'You continue to whittle',
+           'narrowing the connection', 'With a final deep cut')
       waitrt?
     end
     put_in_container('carving knife')
     bput('get shaper', 'You get', 'I could not find', 'What were you referring to')
     while right_hand.include? 'block'
-      fput("shape block to #{@immortal_aspect}")
+      # TODO: Change to bput.  Need complete messaging.  Messages include:
+      # You (clumsily|confidently) begin shaping your
+      # run the shaper over the surface
+      # You work at shaping
+      # you slowly scape and scratch
+      # continue to carefully buff
+      # continue to carve at its surface
+      # continue to shape the surface
+      # You focus (adj) on bringing out the details
+      # you place the last finishing touch
+      fput("shape my block to #{@immortal_aspect}")
       pause 1
       waitrt?
     end
     put_in_container('shaper')
     get_from_container('prayer chain')
-    fput "put #{@immortal_aspect} bead on prayer chain"
-    if worn_chain
-      bput('wear prayer chain', 'You attach', 'You are already wearing that', 'Wear what?')
-    else
-      put_in_container('prayer chain')
-    end
-  end
+    bput("put #{@immortal_aspect} bead on prayer chain",
+         'You carefully thread', 'What were you referring to')
 
-  ######################## DEVOTION RITUALS ########################
-
-  def clean_altar
-    return unless inside?('holy water', @water_holder)
-    get_from_container(@water_holder)
-    fput 'clean altar with holy water'
-    waitfor 'You finish your job'
-    put_in_container(@water_holder)
-    waitrt?
+    replace_prayer_chain
+    true
   end
 
   def meditate_bead
-    return unless @immortal_aspect
-
-    case bput('get prayer chain', 'You get', 'I could not find', 'What were you referring to', 'already in your inventory')
+    get_prayer_chain
+    if count_prayer_beads == 0
+      replace_prayer_chain
+      return false
+    end
+    walk_to @data['altar']['id']
+    fput 'kneel'
+    bput('meditate my prayer chain', 'You clutch')
+    waitfor 'suddenly detaches from your prayer bead chain'
+    fix_standing
+    replace_prayer_chain
+    waitrt?
+    true
+  end
+  
+  def get_prayer_chain
```

Comment:
What do you think about adding `?` to the end of the function name to denote it is a predicate?

Reply:
This is a Ruby naming convention.

I think it might apply to each of the ritual methods too.

Reply:
I'm in favor of this change here.  For the ritual methods I'm on the fence.  I think my hesitation stems from the fact that get_prayer_chain returns false in the case that it hasn't mutated state, while the ritual methods can change the state of the world even though they return false.